src/Common/Scheduler/Workload/WorkloadEntityConfigStorage.cpp

```diff
@@ -0,0 +1,197 @@
+{
+}
+
+void WorkloadEntityConfigStorage::loadEntities()
+{
+    if (entities_loaded.exchange(true))
+        return;
+    
+    // Config entities are loaded via updateConfiguration, not here
+    // Just mark as loaded so subsequent calls don't do anything
+}
+
+void WorkloadEntityConfigStorage::updateConfiguration(const Poco::Util::AbstractConfiguration & config)
+{
+    current_config = &config;
+    refreshEntities(config);
+}
+
+WorkloadEntityConfigStorage::OperationResult WorkloadEntityConfigStorage::storeEntityImpl(
+    const ContextPtr &,
+    WorkloadEntityType,
+    const String &,
+    ASTPtr,
+    bool,
+    bool,
+    const Settings &)
+{
+    // Config storage is read-only - entities come from config, not SQL
+    return OperationResult::Failed;
+}
+
+WorkloadEntityConfigStorage::OperationResult WorkloadEntityConfigStorage::removeEntityImpl(
+    const ContextPtr &,
+    WorkloadEntityType,
+    const String &,
+    bool)
+{
+    // Config storage is read-only - entities come from config, not SQL
+    return OperationResult::Failed;
+}
+
+void WorkloadEntityConfigStorage::refreshEntities(const Poco::Util::AbstractConfiguration & config)
+{
+    LOG_DEBUG(log, "Refreshing workload entities from configuration");
+
+    std::vector<std::pair<String, ASTPtr>> new_entities;
+
+    // Load resources first (workloads may reference them)
+    if (config.has("predefined_resources"))
+    {
+        Poco::Util::AbstractConfiguration::Keys resource_names;
+        config.keys("predefined_resources", resource_names);
+
+        for (const auto & resource_name : resource_names)
+        {
+            String config_key = "predefined_resources." + resource_name;
+            if (!config.has(config_key + ".sql"))
+            {
+                LOG_WARNING(log, "Resource '{}' is missing 'sql' configuration", resource_name);
+                continue;
+            }
+
+            String sql = config.getString(config_key + ".sql");
+            
+            if (auto ast = parseEntityFromConfig(WorkloadEntityType::Resource, resource_name, sql))
+                new_entities.emplace_back(resource_name, ast);
+        }
+    }
+
+    // Load workloads after resources
+    if (config.has("predefined_workloads"))
+    {
+        Poco::Util::AbstractConfiguration::Keys workload_names;
+        config.keys("predefined_workloads", workload_names);
+
+        for (const auto & workload_name : workload_names)
+        {
+            String config_key = "predefined_workloads." + workload_name;
+            if (!config.has(config_key + ".sql"))
+            {
+                LOG_WARNING(log, "Workload '{}' is missing 'sql' configuration", workload_name);
+                continue;
+            }
+
+            String sql = config.getString(config_key + ".sql");
+            
+            if (auto ast = parseEntityFromConfig(WorkloadEntityType::Workload, workload_name, sql))
+                new_entities.emplace_back(workload_name, ast);
+        }
+    }
```

Comment:
Burrow the parsing logic from WorkloadEntityKeeperStorage::refreshEntities. To avoid duplicating it, add a new parsing function as a protected method in WorkloadEntityKeeperStorage and reuse it from both WorkloadEntityConfigStorage and WorkloadEntityKeeperStorage.